### PR TITLE
Proxito: use unresolver in 404 handler

### PR DESCRIPTION
**Pull request by stsewd**
- Moved the logic around checking if an index redirect can be done and finding a custom 404 file into functions, so they can be re-used in both implementations.
- Logic is the same
- One small change is that instead of checking if the version slug is defined now we are checking if we were able to find a version object, otherwise looking for an index file or custom 404 page in a version that doesn't exist will be useless (this doesn't change the logic, just an optimization).

This is on top of https://github.com/readthedocs/readthedocs.org/pull/10037 and requires https://github.com/readthedocs/readthedocs.org/pull/10044. I'll update this PR once those are merged.

### CODE REVIEW

File: readthedocs/proxito/views/serve.py
Comment (benjaoming) [inline readthedocs/proxito/views/serve.py:522]: I have some issues understanding how subprojects are handled here, noting for later :+1:

**Comment by benjaoming** [comment]
@stsewd heads up, I've added some commits here to make the refactor work, I'll take it for a test run and push changes directly unless you're working on it right now?

**Comment by benjaoming** [comment]
The PR currently looks a bit crazy until refactor-unresolver also has merged in the latest `main` :+1:

File: readthedocs/proxito/tests/test_old_redirects.py
Comment (stsewd) [inline readthedocs/proxito/tests/test_old_redirects.py:1150]: We could also make the old implementation compatible, isn't a lot of code, but I'm not sure if we want to spend more time there... (but kind of already spent time debugging it :_)

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [inline readthedocs/proxito/views/serve.py:685]: This is ready for review... but I don't like that we are defining lots of variables that are kind of related, this is kind of a copy/paste of the old implementation, but maybe there is a better way...

Comment (stsewd) [reply readthedocs/proxito/views/serve.py:685]: They are mainly needed for custom redirects, maybe the answer is there... if we can simplify that.

Comment (stsewd) [reply readthedocs/proxito/views/serve.py:685]: And, yeah, we should be able to remove `version_slug` and `lang_slug` with https://github.com/readthedocs/readthedocs.org/issues/10112 :)

**Comment by ericholscher** [comment]
Is this ready to review? How does it related to https://github.com/readthedocs/readthedocs.org/pull/9657?

File: readthedocs/core/unresolver.py
Comment (ericholscher) [inline readthedocs/core/unresolver.py:59]: Should this just be called `version`? Are the other variables here not slugs?

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [inline readthedocs/proxito/views/serve.py:730]: This logic isn't super clear? Why are we only doing this when there is a version? I think the original `get` method had good comments around this stuff -- probably worth copying :)

**Review by ericholscher — COMMENTED** [review]
This is a useful refactor, but want to make sure we aren't losing important context/comments in the refactor. Probably just needs to be compared again current `main`, and adapted from all the previous reviews. 

Really great work knocking down these refactors @stsewd 🎉

File: readthedocs/core/unresolver.py
Comment (stsewd) [reply readthedocs/core/unresolver.py:59]: We are using `project` and `version` in several parts of proxito, so being explicit that this isn't an object, but a slug, makes sense to me. For this particular case, we don't have access to the full version (it doesn't exist), so we just have the slug.

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:466]: The unresolved domain has the project that's attached to that domain and other important information that's used in proxito views

https://github.com/readthedocs/readthedocs.org/blob/96fe6694a85c6e2b5e840fe0600ca7c3ca256a3b/readthedocs/proxito/middleware.py#L186-L192

https://github.com/readthedocs/readthedocs.org/blob/96fe6694a85c6e2b5e840fe0600ca7c3ca256a3b/readthedocs/core/unresolver.py#L105-L128

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:667]: yeah logic is the same, copy/paste, just the variable names changed `final_project -> project`, `version_slug -> version.slug` and so.

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:679]: The version that results from the `unresolve_path()` call already is validated to use the correct manager, this is more like an extra protection (and to reuse the same methods with the older implementation)

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:719]: yeah, the names aren't correct, we sometimes pass the slug from the external version domain or the default version on single version projects. But this is what we expect in the middleware

**Comment by stsewd** [comment]
> Is this ready to review? How does it relate to https://github.com/readthedocs/readthedocs.org/pull/9657?

This is ready, #9657 uses this branch as base, and it can be reviewed/merged separately.

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:716]: Put a comment at the top of the try, since an exception happened, we fill the other variables with the information that we have available.

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [reply readthedocs/proxito/views/serve.py:466]: Hmm. Gotcha. This does make sense, I just haven't fully understood this refactor, I think. Thanks for the links 👍

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [reply readthedocs/proxito/views/serve.py:683]: Cool. Probably worth noting that in the comment as well (I might like comments too much, but mission-critical code is important note this stuff), since I thought this was `SECURITY CRITICAL`, instead of `Be extra defensive`...

```suggestion
        # We force all storage calls to use the external versions storage,
        # since we are serving an external version.
        # The version that results from the unresolve_path() call already is validated to use the correct manager,
        # this is here to add defense in depth against serving the wrong version.
        if unresolved_domain.is_from_external_domain:
            self.version_type = EXTERNAL
```

**Review by ericholscher — APPROVED** [review]
This looks good to merge after my comment suggestions are added in :)

File: readthedocs/core/unresolver.py
Comment (humitos) [reply readthedocs/core/unresolver.py:59]: 👍🏼 to add that comment here.

Also, it makes sense to reorder the arguments to keep consistency with the other exceptions. Usually it comes `project, version_slug`, but this one is different.

File: readthedocs/proxito/tests/test_old_redirects.py
Comment (humitos) [reply readthedocs/proxito/tests/test_old_redirects.py:1150]: "The old implementation", is the one we are going to delete, right?

File: readthedocs/proxito/views/serve.py
Comment (humitos) [inline readthedocs/proxito/views/serve.py:497]: What do you mean by "redirect ot its index file"? `/en/latest/file/` -> `/en/latest/file/index.html`?

If that's correct, can you expand the comment with an example similar to the one I wrote?

File: readthedocs/proxito/views/serve.py
Comment (humitos) [reply readthedocs/proxito/views/serve.py:689]: yeah, I don't like too much the one-line assignment too much because it's harder to read the name of the variables. But that's a personal preference 😄

**Review by humitos — APPROVED** [review]
Looks good to me. I left some small comments, tho.

File: readthedocs/proxito/tests/test_old_redirects.py
Comment (stsewd) [reply readthedocs/proxito/tests/test_old_redirects.py:1150]: yeah

**Pull request merged**
